### DISCUSSION TITLE
No logging for exception in Converter

### DISCUSSION THREAD

**Discussion by llfab** [AvaloniaUI/Avalonia#10618] [Q&A] [answered]
I have created an ```IValueConverter``` where the ```ConvertBack``` function throws an exception like this:

```csharp
public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
{
    throw new NotImplementedException();
}
```

We have integrated into Avalonia tracing so we get all the Avalonia log messages (e.g. false Bindings) into our own tracing. Accedently, I have used above Converter in a two-way binding, so the ```ConvertBack``` was actually called. However, this did not lead to any log message or exception in our code. Hence, I assume that such exceptions are caught within Avalonia. 

Is it expected that no message what so ever is being created?

**Comment by timunie** [comment, marked as answer]
For me it does crashes the App completely. The debugger breaks at the Exception. So it's something I guess you have to log via global exception handling. (Tested using latest nightly)

https://docs.avaloniaui.net/docs/getting-started/unhandledexceptions

**Reply by llfab** [reply]
The debugger does stop if you have set it to halt on all exceptions, yes. But I'm looking for a "notification" in normal runs. We are using tracing heavily and I am not the biggest fan of always use the Debugger...

**Reply by llfab** [reply]
So, I double checked again. Yes, the Debug-Build Debugger would halt, but that does not secure the way for any kind of use, like running it outside debugger or in Release-Build. Also, the Exception does not bubble up to main(), so it must be caught somewhere in Avalonia...

**Reply by llfab** [reply]
This is where it happens:
![image](https://user-images.githubusercontent.com/31031996/224048057-4e36e125-b857-4974-938b-631d3965ce16.png)
(Note that the "wrong" converter is called once data context is applied)

![image](https://user-images.githubusercontent.com/31031996/224048155-d505b08c-089f-4804-82a5-75eda8833b71.png)

... and the Exception is not bubbled up. Setting the data context returns and everything continues...

**Reply by llfab** [reply]
I guess one would need to run Avalonia from source an then see where it is caught...

**Reply by timunie** [reply]
For me the app crashed in Release mode. But if you consider it a bug, please open a bug report and provide a sample to reproduce if possible.

**Reply by llfab** [reply]
Interesting. Then it may be something caught at my end. Let me try with a very simple app. I will post the results. Thanks for the support.

**Reply by llfab** [reply]
Ok, I double checked and on my end the exception in the ConvertBack is thrown, the Debugger halts because of its settings but normal run does not crash or log or whatsoever anything. Please find the sample here: https://github.com/llfab/Samples/tree/main/AvaloniaApplication10

**Reply by timunie** [reply]
I downloaded your sample. When the Binding is applied, yes I see the same as you. But the error is and should get handled by Avalonia imo, as the ConvertBack should not be called at this time. You don't want to have your ViewModel modified by the UI before the user can even interact with it. Do you agree? 

Once the UI is ready, as soon as I click the Button `ConvertBack` is called and I get this output:

![image](https://user-images.githubusercontent.com/47110241/224950219-9658da3b-9159-4924-848d-c5872b5f3c0b.png)

The App closes after that. Tested both in Debug mode and in Release mode (Windows 11 only)

Only line I changed is `Console.WriteLine` ► `Trace.WriteLine`

**Reply by llfab** [reply]
Same here (I did miss though to click :-) after nothing happened on startup. My bad...). Yes, you could argue that while data binding is applied, exceptions should be caught. However, I believe it would be beneficial, if Avalonia logged a ```Warning``` in such a case as it does for other "wrong" bindings. WPF has already made (in my opinion) the mistake to disable WPF tracing unless you are running the app in the Debugger. I find tracing super helpful. And these kind of Xaml-Texts not matching your ViewModel items happens often. I would always prefer to have some notifications even outside the Debugger.

**Reply by timunie** [reply]
That's beyond my knowledge and probably also almost a philosophical question. I still think it's not a bug, but an idea would be to open a feature request. Can you prepare a feature request to extend logging here? Best if you can link your sample there. 

Once community PRs are open again, you can also file a PR if you want to. I cannot guarantee it will be merged, but worth a try imo. 

Happy coding
Tim

**Reply by llfab** [reply]
Agree, tracing vs. debugging is close to religion... and yes it is not a bug. I kind of had expected a warning, again, because for some other aspects the Avalonia log outputs are pretty helpful. => I will do the Feature Request

**Reply by llfab** [reply]
https://github.com/AvaloniaUI/Avalonia/issues/10656
